File: windows/src/Nabira.Win/Core/TechnicalAbbreviations.cs
namespace Nabira.Win.Core;

/// <summary>Technical abbreviations that OS spell checkers frequently do not treat as words.
/// Keep this list in sync with the macOS implementation and the shared test data.</summary>
internal static class TechnicalAbbreviations
{
    internal static readonly IReadOnlySet<string> CanonicalForms = new HashSet<string>
    {
        "API", "ASCII", "AWS", "BIOS", "CDN", "CLI", "CPU", "CRM", "CSS", "CSV",
        "DLL", "DMG", "DNS", "DPI", "FAQ", "FPS", "FTP", "GCP", "GIF", "GPU",
        "GPT", "GUI", "HDD", "HTML", "HTTP", "HTTPS", "IDE", "IMAP", "JPEG", "JPG",
        "JSON", "JWT", "LAN", "LDAP", "LLM", "LTE", "MSI", "NAT", "NFC", "NLP",
        "OAuth", "OCR", "PDF", "PNG", "RDP", "RFID", "SaaS", "SDK", "SFTP", "SMTP",
        "SQL", "SSH", "SSD", "SSL", "TCP", "TSV", "UDP", "URL", "USB", "UUID",
        "VDS", "VPN", "VPS", "WLAN", "XML", "YAML"
    };

    private static readonly IReadOnlyDictionary<string, string> ByLowercase = CanonicalForms
        .ToDictionary(value => value.ToLowerInvariant(), value => value, StringComparer.Ordinal);

    internal static string? CanonicalForm(string word, string language)
    {
        if (!language.StartsWith("en", StringComparison.OrdinalIgnoreCase) ||
            word.Length is < 3 or > 12 ||
            !word.All(c => c is >= 'A' and <= 'Z' or >= 'a' and <= 'z'))
            return null;
        return ByLowercase.GetValueOrDefault(word.ToLowerInvariant());
    }

    /// <summary>Strong positive RU→EN layout signal. It runs before the ALL-CAPS and dictionary
    /// vetoes so <c>мзт</c> can become canonical <c>VPN</c> instead of a Russian typo.</summary>
    internal static string? AutomaticReplacement(
        string typed, string converted, string currentLanguage, string otherLanguage)
    {
        if (!currentLanguage.StartsWith("ru", StringComparison.OrdinalIgnoreCase) ||
            !otherLanguage.StartsWith("en", StringComparison.OrdinalIgnoreCase) ||
            typed.Length != converted.Length ||
            !typed.All(c => c is >= '\u0400' and <= '\u04FF'))
            return null;
        return CanonicalForm(converted, otherLanguage);
    }
}

File: macos/Sources/Nabira/TechnicalAbbreviations.swift
import Foundation

/// Технические сокращения, которые системные словари часто не считают словами.
///
/// Поведенческий контракт и общий список тестовых данных находятся в
/// `shared/docs/technical-abbreviations.md` и `shared/testdata/technical-abbreviations.json`.
/// Swift- и Windows-реализации намеренно отдельные, но обязаны содержать один набор.
enum TechnicalAbbreviations {
    static let canonicalForms: Set<String> = [
        "API", "ASCII", "AWS", "BIOS", "CDN", "CLI", "CPU", "CRM", "CSS", "CSV",
        "DLL", "DMG", "DNS", "DPI", "FAQ", "FPS", "FTP", "GCP", "GIF", "GPU",
        "GPT", "GUI", "HDD", "HTML", "HTTP", "HTTPS", "IDE", "IMAP", "JPEG", "JPG",
        "JSON", "JWT", "LAN", "LDAP", "LLM", "LTE", "MSI", "NAT", "NFC", "NLP",
        "OAuth", "OCR", "PDF", "PNG", "RDP", "RFID", "SaaS", "SDK", "SFTP", "SMTP",
        "SQL", "SSH", "SSD", "SSL", "TCP", "TSV", "UDP", "URL", "USB", "UUID",
        "VDS", "VPN", "VPS", "WLAN", "XML", "YAML",
    ]

    private static let byLowercase: [String: String] = Dictionary(
        uniqueKeysWithValues: canonicalForms.map { ($0.lowercased(), $0) }
    )

    /// Возвращает принятую форму сокращения (`vpn` → `VPN`, `oauth` → `OAuth`).
    static func canonicalForm(for word: String, language: String) -> String? {
        guard String(language.lowercased().prefix(2)) == "en",
              word.count >= 3,
              word.count <= 12,
              word.unicodeScalars.allSatisfy({
                  (0x41...0x5A).contains($0.value) || (0x61...0x7A).contains($0.value)
              }) else { return nil }
        return byLowercase[word.lowercased()]
    }

    /// Сильный положительный сигнал неправильной RU→EN раскладки. Срабатывает до
    /// словаря и вето ALL-CAPS: иначе `мзт` остаётся без layout-кандидата и корректор
    /// русских опечаток успевает заменить его на более частое «мат».
    static func automaticReplacement(
        typed: String,
        converted: String,
        currentLanguage: String,
        otherLanguage: String
    ) -> String? {
        let current = String(currentLanguage.lowercased().prefix(2))
        let other = String(otherLanguage.lowercased().prefix(2))
        guard current == "ru", other == "en",
              typed.count == converted.count,
              typed.unicodeScalars.allSatisfy({ (0x0400...0x04FF).contains($0.value) }) else {
            return nil
        }
        return canonicalForm(for: converted, language: other)
    }
}
